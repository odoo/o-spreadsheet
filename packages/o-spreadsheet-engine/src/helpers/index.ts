export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
