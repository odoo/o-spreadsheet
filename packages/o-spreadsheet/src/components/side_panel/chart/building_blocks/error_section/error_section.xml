<templates>
  <t t-name="o-spreadsheet.ChartErrorSection">
    <Section>
      <ValidationMessages messages="props.messages" msgType="'error'"/>
    </Section>
  </t>
</templates>
