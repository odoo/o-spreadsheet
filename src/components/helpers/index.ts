export * from "@odoo/o-spreadsheet-engine/components/helpers/css";
