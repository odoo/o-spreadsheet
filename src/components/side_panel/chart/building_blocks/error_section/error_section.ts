import { Component } from "@odoo/owl";
import { SpreadsheetChildEnv } from "../../../../../types";
import { ValidationMessages } from "../../../../validation_messages/validation_messages";
import { Section } from "../../../components/section/section";

interface Props {
  messages: string[];
}

export class ChartErrorSection extends Component<Props, SpreadsheetChildEnv> {
  static template = "o-spreadsheet.ChartErrorSection";
  static components = { Section, ValidationMessages };
  static props = { messages: { type: Array, element: String } };
}
