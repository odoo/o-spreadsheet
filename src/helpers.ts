import { Zone } from "./model/state";

/**
 * This files contains a small collection of useful helpers. They are mostly
 * concerned with the following topics:
 *
 * - coordinate conversions
 * - zone manipulation
 * - miscellaneous
 */

//------------------------------------------------------------------------------
// Coordinate
//------------------------------------------------------------------------------

/**
 * Convert a (col) number to the corresponding letter.
 *
 * Examples:
 *     0 => 'A'
 *     25 => 'Z'
 *     26 => 'AA'
 *     27 => 'AB'
 */
export function numberToLetters(n: number): string {
  if (n < 26) {
    return String.fromCharCode(65 + n);
  } else {
    return numberToLetters(Math.floor(n / 26) - 1) + numberToLetters(n % 26);
  }
}

/**
 * Convert a string (describing a column) to its number value.
 *
 * Examples:
 *     'A' => 0
 *     'Z' => 25
 *     'AA' => 26
 */
function lettersToNumber(letters: string): number {
  let result = 0;
  const l = letters.length;
  for (let i = 0; i < l; i++) {
    let n = letters.charCodeAt(i) - 65 + (i < l - 1 ? 1 : 0);
    result += n * 26 ** (l - i - 1);
  }
  return result;
}

/**
 * Convert a "XC" coordinate to cartesian coordinates.
 *
 * Examples:
 *   A1 => [0,0]
 *   B3 => [1,2]
 *
 * Note: it also accepts lowercase coordinates
 */
export function toCartesian(xc: string): [number, number] {
  xc = xc.toUpperCase();
  const [m, letters, numbers] = xc.match(/([A-Z]*)([0-9]*)/)!;
  if (m !== xc) {
    throw new Error(`Invalid cell description: ${xc}`);
  }
  const col = lettersToNumber(letters);
  const row = parseInt(numbers, 10) - 1;
  return [col, row];
}

/**
 * Convert from cartesian coordinate to the "XC" coordinate system.
 *
 * Examples:
 *   [0,0] => A1
 *   [1,2] => B3
 */
export function toXC(col: number, row: number): string {
  return numberToLetters(col) + String(row + 1);
}

export function zoneToXC(zone: Zone): string {
  const topLeft = toXC(zone.left, zone.top);
  const botRight = toXC(zone.right, zone.bottom);

  if (topLeft != botRight) {
    return topLeft + ":" + botRight;
  }

  return topLeft;
}

//------------------------------------------------------------------------------
// Zones
//------------------------------------------------------------------------------

/**
 * Compute the union of two zones. It is the smallest zone which contains the
 * two arguments.
 */
export function union(z1: Zone, z2: Zone): Zone {
  return {
    top: Math.min(z1.top, z2.top),
    left: Math.min(z1.left, z2.left),
    bottom: Math.max(z1.bottom, z2.bottom),
    right: Math.max(z1.right, z2.right)
  };
}

/**
 * Two zones are equal if they represent the same area, so we clearly cannot use
 * reference equality.
 */
export function isEqual(z1: Zone, z2: Zone): boolean {
  return (
    z1.left === z2.left && z1.right === z2.right && z1.top === z2.top && z1.bottom === z2.bottom
  );
}

/**
 * Return true if two zones overlap, false otherwise.
 */
export function overlap(z1: Zone, z2: Zone): boolean {
  if (z1.bottom < z2.top || z2.bottom < z2.top) {
    return false;
  }
  if (z1.right < z2.left || z2.right < z1.left) {
    return false;
  }
  return true;
}

export function isInside(col: number, row: number, zone: Zone): boolean {
  const { left, right, top, bottom } = zone;
  return col >= left && col <= right && row >= top && row <= bottom;
}

//------------------------------------------------------------------------------
// Miscellaneous
//------------------------------------------------------------------------------

/**
 * Stringify an object, like JSON.stringify, except that the first level of keys
 * is ordered.
 */
export function stringify(obj: any): string {
  return JSON.stringify(obj, Object.keys(obj).sort());
}
