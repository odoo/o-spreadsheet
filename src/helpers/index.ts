export * from "./color";
export * from "./coordinates";
export * from "./dates";
export * from "./edge_scrolling";
export * from "./format";
export * from "./misc";
export * from "./numbers";
export * from "./range";
export * from "./references";
export * from "./search";
export * from "./sheet";
export * from "./uuid";
export * from "./zones";
