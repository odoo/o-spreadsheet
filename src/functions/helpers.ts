// HELPERS
import { numberToJsDate, parseDateTime } from "../helpers/dates";
import { isNumber, parseNumber } from "../helpers/numbers";
import { _t } from "../translation";
import { ArgValue, CellValue, Locale, Matrix, Maybe, ValueAndFormat, isMatrix } from "../types";
import { EvaluationError } from "../types/errors";

const SORT_TYPES_ORDER = ["number", "string", "boolean", "undefined"];

export function assert(condition: () => boolean, message: string): void {
  if (!condition()) {
    throw new Error(message);
  }
}
// -----------------------------------------------------------------------------
// FORMAT FUNCTIONS
// -----------------------------------------------------------------------------

const expectNumberValueError = (value: string) =>
  _t(
    "The function [[FUNCTION_NAME]] expects a number value, but '%s' is a string, and cannot be coerced to a number.",
    value
  );

export const expectNumberRangeError = (lowerBound: number, upperBound: number, value: number) =>
  _t(
    "The function [[FUNCTION_NAME]] expects a number value between %s and %s inclusive, but receives %s.",
    lowerBound.toString(),
    upperBound.toString(),
    value.toString()
  );

export const expectStringSetError = (stringSet: string[], value: string) => {
  const stringSetString = stringSet.map((str) => `'${str}'`).join(", ");
  return _t(
    "The function [[FUNCTION_NAME]] has an argument with value '%s'. It should be one of: %s.",
    value,
    stringSetString
  );
};

export function toNumber(value: CellValue | undefined, locale: Locale): number {
  switch (typeof value) {
    case "number":
      return value;
    case "boolean":
      return value ? 1 : 0;
    case "string":
      if (isNumber(value, locale) || value === "") {
        return parseNumber(value, locale);
      }
      const internalDate = parseDateTime(value, locale);
      if (internalDate) {
        return internalDate.value;
      }
      throw new Error(expectNumberValueError(value));
    default:
      if (value instanceof EvaluationError) {
        throw value;
      }
      return 0;
  }
}

export function strictToNumber(value: CellValue | undefined, locale: Locale): number {
  if (value === "") {
    throw new Error(expectNumberValueError(value));
  }
  return toNumber(value, locale);
}

export function toInteger(value: CellValue | undefined, locale: Locale) {
  return Math.trunc(toNumber(value, locale));
}

export function strictToInteger(value: CellValue | undefined, locale: Locale) {
  return Math.trunc(strictToNumber(value, locale));
}

export function assertNumberGreaterThanOrEqualToOne(value: number) {
  assert(
    () => value >= 1,
    _t(
      "The function [[FUNCTION_NAME]] expects a number value to be greater than or equal to 1, but receives %s.",
      value.toString()
    )
  );
}

export function toString(value: CellValue | undefined): string {
  switch (typeof value) {
    case "string":
      return value;
    case "number":
      return value.toString();
    case "boolean":
      return value ? "TRUE" : "FALSE";
    default:
      if (value instanceof EvaluationError) {
        throw value;
      }
      return "";
  }
}

/**
 * Normalize range by setting all the string in the range to lowercase and replacing
 * accent letters with plain letters
 */
export function normalizeRange<T>(range: T[]) {
  return range.map((item) => (typeof item === "string" ? normalizeString(item) : item));
}

/** Normalize string by setting it to lowercase and replacing accent letters with plain letters */
export function normalizeString(str: string) {
  return str
    .toLowerCase()
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "");
}

const expectBooleanValueError = (value: string) =>
  _t(
    "The function [[FUNCTION_NAME]] expects a boolean value, but '%s' is a text, and cannot be coerced to a number.",
    value
  );

export function toBoolean(value: CellValue | undefined): boolean {
  switch (typeof value) {
    case "boolean":
      return value;
    case "string":
      if (value) {
        let uppercaseVal = value.toUpperCase();
        if (uppercaseVal === "TRUE") {
          return true;
        }
        if (uppercaseVal === "FALSE") {
          return false;
        }
        throw new Error(expectBooleanValueError(value));
      } else {
        return false;
      }
    case "number":
      return value ? true : false;
    default:
      if (value instanceof EvaluationError) {
        throw value;
      }
      return false;
  }
}

function strictToBoolean(value: CellValue | undefined): boolean {
  if (value === "") {
    throw new Error(expectBooleanValueError(value));
  }
  return toBoolean(value);
}

export function toJsDate(value: CellValue | undefined, locale: Locale): Date {
  return numberToJsDate(toNumber(value, locale));
}

// -----------------------------------------------------------------------------
// VISIT FUNCTIONS
// -----------------------------------------------------------------------------
function visitArgs<T extends ValueAndFormat | CellValue>(
  args: (T | Matrix<T> | undefined)[],
  cellCb: (a: T) => void,
  dataCb: (a: T | undefined) => void
): void {
  for (let arg of args) {
    if (isMatrix(arg)) {
      // arg is ref to a Cell/Range
      const lenRow = arg.length;
      const lenCol = arg[0].length;
      for (let y = 0; y < lenCol; y++) {
        for (let x = 0; x < lenRow; x++) {
          cellCb(arg[x][y]);
        }
      }
    } else {
      // arg is set directly in the formula function
      dataCb(arg);
    }
  }
}

export function visitAny<T extends ValueAndFormat | CellValue>(
  args: (T | Matrix<T> | undefined)[],
  cb: (a: T) => void
): void {
  visitArgs(args, cb, cb);
}

export function visitNumbers(args: ArgValue[], cb: (arg: number) => void, locale: Locale): void {
  visitArgs(
    args,
    (cellValue) => {
      if (typeof cellValue === "number") {
        cb(cellValue);
      }
    },
    (argValue) => {
      cb(strictToNumber(argValue, locale));
    }
  );
}

// -----------------------------------------------------------------------------
// REDUCE FUNCTIONS
// -----------------------------------------------------------------------------

function reduceArgs<T, M>(
  args: (T | Matrix<T>)[],
  cellCb: (acc: M, a: T) => M,
  dataCb: (acc: M, a: T) => M,
  initialValue: M,
  dir: "rowFirst" | "colFirst" = "rowFirst"
): M {
  let val = initialValue;
  for (let arg of args) {
    if (isMatrix(arg)) {
      // arg is ref to a Cell/Range
      const numberOfCols = arg.length;
      const numberOfRows = arg[0].length;

      if (dir === "rowFirst") {
        for (let row = 0; row < numberOfRows; row++) {
          for (let col = 0; col < numberOfCols; col++) {
            val = cellCb(val, arg[col][row]);
          }
        }
      } else {
        for (let col = 0; col < numberOfCols; col++) {
          for (let row = 0; row < numberOfRows; row++) {
            val = cellCb(val, arg[col][row]);
          }
        }
      }
    } else {
      // arg is set directly in the formula function
      val = dataCb(val, arg);
    }
  }
  return val;
}

export function reduceAny<T, M>(
  args: (T | Matrix<T>)[],
  cb: (acc: M, a: T) => M,
  initialValue: M,
  dir: "rowFirst" | "colFirst" = "rowFirst"
): M {
  return reduceArgs(args, cb, cb, initialValue, dir);
}

export function reduceNumbers(
  args: ArgValue[],
  cb: (acc: number, a: number) => number,
  initialValue: number,
  locale: Locale
): number {
  return reduceArgs(
    args,
    (acc, ArgValue) => {
      if (typeof ArgValue === "number") {
        return cb(acc, ArgValue);
      } else if (ArgValue instanceof EvaluationError) {
        throw ArgValue;
      }
      return acc;
    },
    (acc, argValue) => {
      return cb(acc, strictToNumber(argValue, locale));
    },
    initialValue
  );
}

export function reduceNumbersTextAs0(
  args: ArgValue[],
  cb: (acc: number, a: number) => number,
  initialValue: number,
  locale: Locale
): number {
  return reduceArgs(
    args,
    (acc, ArgValue) => {
      if (ArgValue !== undefined && ArgValue !== null) {
        if (typeof ArgValue === "number") {
          return cb(acc, ArgValue);
        } else if (typeof ArgValue === "boolean") {
          return cb(acc, toNumber(ArgValue, locale));
        } else {
          return cb(acc, 0);
        }
      }
      return acc;
    },
    (acc, argValue) => {
      return cb(acc, toNumber(argValue, locale));
    },
    initialValue
  );
}

// -----------------------------------------------------------------------------
// MATRIX FUNCTIONS
// -----------------------------------------------------------------------------

/**
 * Generate a matrix of size nColumns x nRows and apply a callback on each position
 */
export function generateMatrix<T>(
  nColumns: number,
  nRows: number,
  callback: (col: number, row: number) => T
): Matrix<T> {
  const returned = Array(nColumns);
  for (let col = 0; col < nColumns; col++) {
    returned[col] = Array(nRows);
    for (let row = 0; row < nRows; row++) {
      returned[col][row] = callback(col, row);
    }
  }
  return returned;
}

export function matrixMap<T, M>(matrix: Matrix<T>, fn: (value: T) => M): Matrix<M> {
  if (matrix.length === 0) {
    return [];
  }
  return generateMatrix(matrix.length, matrix[0].length, (col, row) => fn(matrix[col][row]));
}

export function transposeMatrix<T>(matrix: Matrix<T>): Matrix<T> {
  if (!matrix.length) {
    return [];
  }
  return generateMatrix(matrix[0].length, matrix.length, (i, j) => matrix[j][i]);
}

// -----------------------------------------------------------------------------
// CONDITIONAL EXPLORE FUNCTIONS
// -----------------------------------------------------------------------------

/**
 * This function allows to visit arguments and stop the visit if necessary.
 * It is mainly used to bypass argument evaluation for functions like OR or AND.
 */
function conditionalVisitArgs(
  args: ArgValue[],
  cellCb: (a: CellValue | undefined) => boolean,
  dataCb: (a: Maybe<CellValue>) => boolean
): void {
  for (let arg of args) {
    if (isMatrix(arg)) {
      // arg is ref to a Cell/Range
      const lenRow = arg.length;
      const lenCol = arg[0].length;
      for (let y = 0; y < lenCol; y++) {
        for (let x = 0; x < lenRow; x++) {
          if (!cellCb(arg[x][y] ?? undefined)) return;
        }
      }
    } else {
      // arg is set directly in the formula function
      if (!dataCb(arg)) return;
    }
  }
}

export function conditionalVisitBoolean(args: ArgValue[], cb: (a: boolean) => boolean): void {
  return conditionalVisitArgs(
    args,
    (ArgValue) => {
      if (typeof ArgValue === "boolean") {
        return cb(ArgValue);
      }
      if (typeof ArgValue === "number") {
        return cb(ArgValue ? true : false);
      }
      return true;
    },
    (argValue) => {
      if (argValue !== undefined && argValue !== null) {
        return cb(strictToBoolean(argValue));
      }
      return true;
    }
  );
}

// -----------------------------------------------------------------------------
// CRITERION FUNCTIONS
// -----------------------------------------------------------------------------

type Operator = ">" | ">=" | "<" | "<=" | "<>" | "=";
interface Predicate {
  operator: Operator;
  operand: number | string | boolean;
  regexp?: RegExp;
}

function getPredicate(descr: string, isQuery: boolean, locale: Locale): Predicate {
  let operator: Operator;
  let operand: Maybe<CellValue>;

  let subString = descr.substring(0, 2);

  if (subString === "<=" || subString === ">=" || subString === "<>") {
    operator = subString;
    operand = descr.substring(2);
  } else {
    subString = descr.substring(0, 1);
    if (subString === "<" || subString === ">" || subString === "=") {
      operator = subString;
      operand = descr.substring(1);
    } else {
      operator = "=";
      operand = descr;
    }
  }

  if (isNumber(operand, locale)) {
    operand = toNumber(operand, locale);
  } else if (operand === "TRUE" || operand === "FALSE") {
    operand = toBoolean(operand);
  }

  const result: Predicate = { operator, operand };

  if (typeof operand === "string") {
    if (isQuery) {
      operand += "*";
    }
    result.regexp = operandToRegExp(operand);
  }

  return result;
}

function operandToRegExp(operand: string): RegExp {
  let exp = "";
  let predecessor = "";
  for (let char of operand) {
    if (char === "?" && predecessor !== "~") {
      exp += ".";
    } else if (char === "*" && predecessor !== "~") {
      exp += ".*";
    } else {
      if (char === "*" || char === "?") {
        //remove "~"
        exp = exp.slice(0, -1);
      }
      if (["^", ".", "[", "]", "$", "(", ")", "*", "+", "?", "|", "{", "}", "\\"].includes(char)) {
        exp += "\\";
      }
      exp += char;
    }
    predecessor = char;
  }
  return new RegExp("^" + exp + "$", "i");
}

function evaluatePredicate(value: CellValue | undefined, criterion: Predicate): boolean {
  const { operator, operand } = criterion;

  if (value === undefined || operand === undefined || value === null || operand === null) {
    return false;
  }

  if (typeof operand === "number" && operator === "=") {
    return toString(value) === toString(operand);
  }

  if (operator === "<>" || operator === "=") {
    let result: boolean;
    if (typeof value === typeof operand) {
      if (typeof value === "string" && criterion.regexp) {
        result = criterion.regexp.test(value);
      } else {
        result = value === operand;
      }
    } else {
      result = false;
    }
    return operator === "=" ? result : !result;
  }

  if (typeof value === typeof operand) {
    switch (operator) {
      case "<":
        return value < operand;
      case ">":
        return value > operand;
      case "<=":
        return value <= operand;
      case ">=":
        return value >= operand;
    }
  }
  return false;
}

/**
 * Functions used especially for predicate evaluation on ranges.
 *
 * Take ranges with same dimensions and take predicates, one for each range.
 * For (i, j) coordinates, if all elements with coordinates (i, j) of each
 * range correspond to the associated predicate, then the function uses a callback
 * function with the parameters "i" and "j".
 *
 * Syntax:
 * visitMatchingRanges([range1, predicate1, range2, predicate2, ...], cb(i,j), likeSelection)
 *
 * - range1 (range): The range to check against predicate1.
 * - predicate1 (string): The pattern or test to apply to range1.
 * - range2: (range, repeatable) ranges to check.
 * - predicate2 (string, repeatable): Additional pattern or test to apply to range2.
 *
 * - cb(i: number, j: number) => void: the callback function.
 *
 * - isQuery (boolean) indicates if the comparison with a string should be done as a SQL-like query.
 * (Ex1 isQuery = true, predicate = "abc", element = "abcde": predicate match the element),
 * (Ex2 isQuery = false, predicate = "abc", element = "abcde": predicate not match the element).
 * (Ex3 isQuery = true, predicate = "abc", element = "abc": predicate match the element),
 * (Ex4 isQuery = false, predicate = "abc", element = "abc": predicate match the element).
 */
export function visitMatchingRanges(
  args: ArgValue[],
  cb: (i: number, j: number) => void,
  locale: Locale,
  isQuery: boolean = false
): void {
  const countArg = args.length;

  if (countArg % 2 === 1) {
    throw new Error(
      _t(`Function [[FUNCTION_NAME]] expects criteria_range and criterion to be in pairs.`)
    );
  }

  const dimRow = (args[0] as Matrix<CellValue>).length;
  const dimCol = (args[0] as Matrix<CellValue>)[0].length;

  let predicates: Predicate[] = [];

  for (let i = 0; i < countArg - 1; i += 2) {
    const criteriaRange = args[i];

    if (
      !isMatrix(criteriaRange) ||
      criteriaRange.length !== dimRow ||
      criteriaRange[0].length !== dimCol
    ) {
      throw new Error(
        _t(`Function [[FUNCTION_NAME]] expects criteria_range to have the same dimension`)
      );
    }

    const description = toString(args[i + 1] as Maybe<CellValue>);
    predicates.push(getPredicate(description, isQuery, locale));
  }

  for (let i = 0; i < dimRow; i++) {
    for (let j = 0; j < dimCol; j++) {
      let validatedPredicates = true;
      for (let k = 0; k < countArg - 1; k += 2) {
        const criteriaValue = (args[k] as Matrix<CellValue>)[i][j];
        const criterion = predicates[k / 2];
        validatedPredicates = evaluatePredicate(criteriaValue ?? undefined, criterion);
        if (!validatedPredicates) {
          break;
        }
      }
      if (validatedPredicates) {
        cb(i, j);
      }
    }
  }
}

// -----------------------------------------------------------------------------
// COMMON FUNCTIONS
// -----------------------------------------------------------------------------

/**
 * Perform a dichotomic search on an array and return the index of the nearest match.
 *
 * The array should be sorted, if not an incorrect value might be returned. In the case where multiple
 * element of the array match the target, the method will return the first match if the array is sorted
 * in descending order, and the last match if the array is in ascending order.
 *
 *
 * @param data the array in which to search.
 * @param target the value to search.
 * @param mode "nextGreater/nextSmaller" : return next greater/smaller value if no exact match is found.
 * @param sortOrder whether the array is sorted in ascending or descending order.
 * @param rangeLength the number of elements to consider in the search array.
 * @param getValueInData function returning the element at index i in the search array.
 */
export function dichotomicSearch<T>(
  data: T,
  target: Maybe<CellValue>,
  mode: "nextGreater" | "nextSmaller" | "strict",
  sortOrder: "asc" | "desc",
  rangeLength: number,
  getValueInData: (range: T, index: number) => CellValue | undefined
): number {
  if (target === null || target === undefined) {
    return -1;
  }
  const _target = normalizeValue(target);
  const targetType = typeof _target;

  let matchVal: CellValue | undefined = undefined;
  let matchValIndex: number | undefined = undefined;

  let indexLeft = 0;
  let indexRight = rangeLength - 1;

  let indexMedian: number;
  let currentIndex: number;
  let currentVal: CellValue | undefined;
  let currentType: string;

  while (indexRight - indexLeft >= 0) {
    indexMedian = Math.floor((indexLeft + indexRight) / 2);

    currentIndex = indexMedian;
    currentVal = normalizeValue(getValueInData(data, currentIndex));
    currentType = typeof currentVal;

    // 1 - linear search to find value with the same type
    while (indexLeft < currentIndex && targetType !== currentType) {
      currentIndex--;
      currentVal = normalizeValue(getValueInData(data, currentIndex));
      currentType = typeof currentVal;
    }
    if (currentType !== targetType || currentVal === undefined || currentVal === null) {
      indexLeft = indexMedian + 1;
      continue;
    }

    // 2 - check if value match
    if (mode === "strict" && currentVal === _target) {
      matchVal = currentVal;
      matchValIndex = currentIndex;
    } else if (mode === "nextSmaller" && currentVal <= _target) {
      if (
        matchVal === undefined ||
        matchVal === null ||
        matchVal < currentVal ||
        (matchVal === currentVal && sortOrder === "asc" && matchValIndex! < currentIndex) ||
        (matchVal === currentVal && sortOrder === "desc" && matchValIndex! > currentIndex)
      ) {
        matchVal = currentVal;
        matchValIndex = currentIndex;
      }
    } else if (mode === "nextGreater" && currentVal >= _target) {
      if (
        matchVal === undefined ||
        matchVal > currentVal ||
        (matchVal === currentVal && sortOrder === "asc" && matchValIndex! < currentIndex) ||
        (matchVal === currentVal && sortOrder === "desc" && matchValIndex! > currentIndex)
      ) {
        matchVal = currentVal;
        matchValIndex = currentIndex;
      }
    }

    // 3 - give new indexes for the Binary search
    if (
      (sortOrder === "asc" && currentVal > _target) ||
      (sortOrder === "desc" && currentVal <= _target)
    ) {
      indexRight = currentIndex - 1;
    } else {
      indexLeft = indexMedian + 1;
    }
  }

  // note that valMinIndex could be 0
  return matchValIndex !== undefined ? matchValIndex : -1;
}

/**
 * Perform a linear search and return the index of the match.
 * -1 is returned if no value is found.
 *
 * Example:
 * - [3, 6, 10], 3 => 0
 * - [3, 6, 10], 6 => 1
 * - [3, 6, 10], 9 => -1
 * - [3, 6, 10], 2 => -1
 *
 * @param data the array to search in.
 * @param target the value to search in the array.
 * @param mode if "strict" return exact match index. "nextGreater" returns the next greater
 * element from the target and "nextSmaller" the next smaller
 * @param numberOfValues the number of elements to consider in the search array.
 * @param getValueInData function returning the element at index i in the search array.
 * @param reverseSearch if true, search in the array starting from the end.

 */
export function linearSearch<T>(
  data: T,
  target: Maybe<CellValue> | undefined,
  mode: "nextSmaller" | "nextGreater" | "strict",
  numberOfValues: number,
  getValueInData: (data: T, index: number) => CellValue | undefined,
  reverseSearch = false
): number {
  if (target === null || target === undefined) return -1;

  const _target = normalizeValue(target);
  const getValue = reverseSearch
    ? (data: T, i: number) => getValueInData(data, numberOfValues - i - 1)
    : getValueInData;

  let closestMatch: CellValue | undefined = undefined;
  let closestMatchIndex = -1;
  for (let i = 0; i < numberOfValues; i++) {
    const value = normalizeValue(getValue(data, i));
    if (value === _target) {
      return reverseSearch ? numberOfValues - i - 1 : i;
    }
    if (mode === "nextSmaller") {
      if (
        (!closestMatch && compareCellValues(_target, value) >= 0) ||
        (compareCellValues(_target, value) >= 0 && compareCellValues(value, closestMatch) > 0)
      ) {
        closestMatch = value;
        closestMatchIndex = i;
      }
    } else if (mode === "nextGreater") {
      if (
        (!closestMatch && compareCellValues(_target, value) <= 0) ||
        (compareCellValues(_target, value) <= 0 && compareCellValues(value, closestMatch) < 0)
      ) {
        closestMatch = value;
        closestMatchIndex = i;
      }
    }
  }

  return reverseSearch ? numberOfValues - closestMatchIndex - 1 : closestMatchIndex;
}

/**
 * Normalize a value.
 * If the cell value is a string, this will set it to lowercase and replacing accent letters with plain letters
 */
function normalizeValue<T>(value: T): T | string {
  return typeof value === "string" ? normalizeString(value) : value;
}

function compareCellValues(left: CellValue | undefined, right: CellValue | undefined): number {
  let typeOrder = SORT_TYPES_ORDER.indexOf(typeof left) - SORT_TYPES_ORDER.indexOf(typeof right);
  if (typeOrder === 0) {
    if (typeof left === "string" && typeof right === "string") {
      typeOrder = left.localeCompare(right);
    } else if (typeof left === "number" && typeof right === "number") {
      typeOrder = left - right;
    } else if (typeof left === "boolean" && typeof right === "boolean") {
      typeOrder = Number(left) - Number(right);
    }
  }
  return typeOrder;
}

export function toMatrix<T>(data: T | Matrix<T> | undefined): Matrix<T> {
  if (data === undefined) {
    return [[]];
  }
  return isMatrix(data) ? data : [[data]];
}

/**
 * Flatten an array of items, where each item can be a single value or a 2D array, and apply the
 * callback to each element.
 *
 * The 2D array are flattened row first.
 */
export function flattenRowFirst<T, K>(items: Array<T | Matrix<T>>, callback: (val: T) => K): K[] {
  /**/
  return reduceAny(
    items,
    (array: K[], val: T) => {
      array.push(callback(val));
      return array;
    },
    [],
    "rowFirst"
  );
}
