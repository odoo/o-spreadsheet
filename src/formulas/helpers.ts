import { functionRegistry } from "../functions";
import { AST, ASTFuncall, iterateAstNodes, parseTokens } from "./parser";
import { Token } from "./tokenizer";

const functions = functionRegistry.content;

export function isExportableToExcel(tokens: Token[]): boolean {
  try {
    const nonExportableFunctions = iterateAstNodes(parseTokens(tokens)).filter(
      (ast) => ast.type === "FUNCALL" && !functions[ast.value.toUpperCase()]?.isExported
    );
    return nonExportableFunctions.length === 0;
  } catch (error) {
    return false;
  }
}

export function getFunctionsFromTokens(tokens: Token[], functionNames: string[]) {
  // Parsing is an expensive operation, so we first check if the
  // formula contains one of the function names
  if (!tokens.some((t) => t.type === "SYMBOL" && functionNames.includes(t.value.toUpperCase()))) {
    return [];
  }
  let ast: AST | undefined;
  try {
    ast = parseTokens(tokens);
  } catch {
    return [];
  }
  return getFunctionsFromAST(ast, functionNames);
}

function getFunctionsFromAST(ast: AST, functionNames: string[]) {
  return iterateAstNodes(ast)
    .filter((node) => node.type === "FUNCALL" && functionNames.includes(node.value.toUpperCase()))
    .map((node: ASTFuncall) => ({
      functionName: node.value.toUpperCase(),
      args: node.args,
    }));
}
