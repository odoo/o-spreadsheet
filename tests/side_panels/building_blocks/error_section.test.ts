import { ChartErrorSection } from "../../../src/components/side_panel/chart/building_blocks/error_section/error_section";
import { mountComponent } from "../../test_helpers/helpers";

let fixture: HTMLElement;

async function mountChartErrorSection(props: ChartErrorSection["props"]) {
  ({ fixture } = await mountComponent(ChartErrorSection, { props }));
}

describe("Chart error section", () => {
  test("Can render a chart error section component", async () => {
    await mountChartErrorSection({
      messages: ["error_1", "error_2"],
    });
    expect(fixture).toMatchSnapshot();
  });

  test("Error section does not have a title", async () => {
    await mountChartErrorSection({
      messages: ["error_1", "error_2"],
    });
    expect(fixture.querySelector(".o-section-title")).toBeNull();
  });

  test("Messages are in error", async () => {
    await mountChartErrorSection({
      messages: ["error_1", "error_2"],
    });
    expect(fixture.querySelectorAll(".o-validation-error")).toHaveLength(2);
  });
});
