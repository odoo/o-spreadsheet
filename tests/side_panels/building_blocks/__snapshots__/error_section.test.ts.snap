// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart error section Can render a chart error section component 1`] = `
<div>
  <div
    class="o-section"
  >
    
    <div
      class="d-flex align-items-center o-side-panel-error o-validation-error text-danger"
    >
      <div
        class="o-icon fa-small"
      >
        <i
          class="fa fa-exclamation-triangle"
        />
      </div>
      
      <span>
        error_1
      </span>
    </div>
    <div
      class="d-flex align-items-center o-side-panel-error o-validation-error text-danger"
    >
      <div
        class="o-icon fa-small"
      >
        <i
          class="fa fa-exclamation-triangle"
        />
      </div>
      
      <span>
        error_2
      </span>
    </div>
    
    
  </div>
</div>
`;
